sql/plan/recursive_cte.go

```diff
@@ -0,0 +1,354 @@
+	return r.name
+}
+
+// WithSchema inherits [Init]'s schema at resolve time
+func (r *RecursiveCte) WithSchema(s sql.Schema) *RecursiveCte {
+	nr := *r
+	nr.schema = s
+	return &nr
+}
+
+// WithWorking populates the [working] table with a common schema
+func (r *RecursiveCte) WithWorking(t *RecursiveTable) *RecursiveCte {
+	nr := *r
+	nr.working = t
+	return &nr
+}
+
+// Schema implements sql.Node
+func (r *RecursiveCte) Schema() sql.Schema {
+	return r.schema
+}
+
+// RowIter implements sql.Node
+func (r *RecursiveCte) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	return &recursiveCteIter{
+		init:        r.Init,
+		rec:         r.Rec,
+		row:         row,
+		working:     r.working,
+		temp:        make([]sql.Row, 0),
+		deduplicate: r.Deduplicate,
+	}, nil
+}
+
+// Children implements sql.Node
+func (r *RecursiveCte) Children() []sql.Node {
+	return []sql.Node{r.Init, r.Rec}
+}
+
+// WithChildren implements sql.Node
+func (r *RecursiveCte) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(r, len(children), 2)
+	}
+	nn := *r
+	nn.Init = children[0]
+	nn.Rec = children[1]
+	return &nn, nil
+}
+
+// Resolved implements sql.Node
+func (r *RecursiveCte) Resolved() bool {
+	return r.Init.Resolved() && r.Rec.Resolved()
+}
+
+// String implements sql.Node
+func (r *RecursiveCte) String() string {
+	var union string
+	if r.Deduplicate {
+		union = "UNION"
+	} else {
+		union = "UNION ALL"
+	}
+	return fmt.Sprintf("(%s %s %s)", r.Init, union, r.Rec)
+}
+
+// Type implements sql.Node
+func (r *RecursiveCte) Type() sql.Type {
+	cols := r.schema
+	if len(cols) == 1 {
+		return cols[0].Type
+	}
+	ts := make([]sql.Type, len(cols))
+	for i, c := range cols {
+		ts[i] = c.Type
+	}
+	return sql.CreateTuple(ts...)
+}
+
+// IsNullable implements sql.Node
+func (r *RecursiveCte) IsNullable() bool {
+	return true
+}
+
+// recursiveCteIter exhaustively executes a recursive
+// relation [rec] populated by an [init] base case.
+// Refer to RecursiveCte for more details.
+type recursiveCteIter struct {
+	// base sql.Project
+	init sql.Node
+	// recursive sql.Project
+	rec sql.Node
+	// anchor to recursive table to repopulate with [temp]
+	working *RecursiveTable
+	// true if UNION, false if UNION ALL
+	deduplicate bool
+	// parent iter initialization state
+	row sql.Row
+
+	// active iterator, either [init].RowIter or [rec].RowIter
+	iter sql.RowIter
+	// number of recursive iterations finished
+	cycle int
+	// buffer to collect intermediate results for next recursion
+	temp []sql.Row
+	// duplicate lookup if [deduplicated] set
+	cache map[uint64]struct{}
+}
+
+var _ sql.RowIter = (*recursiveCteIter)(nil)
+
+// Next implements sql.RowIter
+func (r *recursiveCteIter) Next(ctx *sql.Context) (sql.Row, error) {
+	if r.iter == nil {
+		// start with [Init].RowIter
+		var err error
+		if r.deduplicate {
+			r.cache = make(map[uint64]struct{})
```

Comment:
Should use the cache methods for this instead (so that when we finally add memory management we'll have fewer places to change)